li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -929,4 +1011,109 @@ private void refreshSubscriptionMetadata() {
     }
   }
+
+  // Callback interface for concurrently calling a method against per-cluster consumers with cluster-specific input
+  // values with timeout.
+  interface PerClusterExecutorCallback<K, V, I> {
+    void execute(LiKafkaConsumer<K, V> consumer, I perClusterInput, Duration timeout);
+  }
+
+  // For each cluster in the perClusterInput key set, concurrently call the given callback against the corresponding
+  // consumer with the input values for that cluster. If perClusterInput is null, the callback will be called for all
+  // existing consumers.
+  private <I> void executePerClusterCallback(String methodName,
+      Map<ClusterDescriptor, I> perClusterInput, PerClusterExecutorCallback callback, Duration timeout) {
+    List<ClusterConsumerPair<K, V>> consumers;
+    if (perClusterInput == null) {
+      // The callback needs to be executed against all consumers
+      consumers = _consumers;
+    } else {
+      consumers = new ArrayList<>();
+      for (ClusterDescriptor cluster : perClusterInput.keySet()) {
+        LiKafkaConsumer<K, V> consumer = getPerClusterConsumer(cluster);
+        if (cluster == null) {
+          LOG.warn("consumer for cluster {} does not exist for {}. ignoring.", cluster, methodName);
+        } else {
+          consumers.add(new ClusterConsumerPair<K, V>(cluster, consumer));
+        }
+      }
+    }
+
+    if (consumers.isEmpty()) {
+      // Nothing to do
+      return;
+    }
+
+    LOG.debug("starting {} for {} LiKafkaConsumers for cluster group {} in {} {}", methodName, consumers.size(),
+        _clusterGroup, timeout);
+
+    long startTimeMs = System.currentTimeMillis();
+    long deadlineTimeMs = startTimeMs + timeout.toMillis();
+    CountDownLatch countDownLatch = new CountDownLatch(consumers.size());
+    Set<Thread> threads = new HashSet<>();
+    for (ClusterConsumerPair<K, V> entry : consumers) {
+      ClusterDescriptor cluster = entry.getCluster();
+      LiKafkaConsumer<K, V> consumer = entry.getConsumer();
+      Thread t = new Thread(() -> {
+        try {
+          Duration remainingTime = Duration.ofMillis(Math.max(deadlineTimeMs - System.currentTimeMillis(), 0));
+          callback.execute(consumer, perClusterInput == null ? null : perClusterInput.get(cluster), remainingTime);
+        } finally {
+          countDownLatch.countDown();
+        }
+      });
+      t.setDaemon(true);
+      t.setName("LiKafkaConsumer-" + methodName + "-" + cluster.getName());
+      t.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+        public void uncaughtException(Thread t, Throwable e) {
+          throw new KafkaException("Thread " + t.getName() + " throws exception", e);
+        }
+      });
+      t.start();
+      threads.add(t);
+    }
+
+    try {
+      if (!countDownLatch.await(deadlineTimeMs - System.currentTimeMillis(), TimeUnit.MILLISECONDS)) {
```

Comment:
Same question here, not sure if we need to consider negative values for await here.